File: apps/web/src/types/tile.ts
export type Position = [number, number];

export type TileKind = 'Empty' | 'Wall' | 'SoftWall' | 'Entrypoint' | 'Checkpoint' | 'Exit';

export type TileHighlight = {
  significancy: 1 | 2 | 3 | 4 | 5 | 6 | 7;
};

export type Tile =
  | { kind: Exclude<TileKind, 'Checkpoint'>; highlighted?: TileHighlight }
  | {
      kind: Extract<TileKind, 'Checkpoint'>;
      level: number;
      highlighted?: TileHighlight;
    };

export type MazeTileBoardBuilder = Tile[][];
export type MazeTileBoard = readonly (readonly Tile[])[];

File: crates/engine-core/src/tile.rs
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TileKind {
    Entrypoint,
    Empty,
    Wall,
    Checkpoint { level: i32 },
}

pub type TileBoard = Vec<Vec<TileKind>>;

#[derive(Clone, Copy, PartialEq, Debug)]
pub struct Position {
    pub x: usize,
    pub y: usize,
}

#[derive(Clone, Copy, PartialEq, Debug)]
pub struct Checkpoint {
    pub position: Position,
    pub level: i32,
}
